radix-engine-interface/src/data/custom_schema.rs

```diff
@@ -141,6 +143,106 @@ impl CustomTypeExtension for ScryptoCustomTypeExtension {
+    ) -> Result<(), SchemaValidationError> {
+        // Even though they all map to the same thing, we keep the explicit match statement so that
+        // we will have to explicitly check this when we add a new `ScryptoCustomTypeKind`
+        match type_kind {
+            ScryptoCustomTypeKind::PackageAddress
+            | ScryptoCustomTypeKind::ComponentAddress
+            | ScryptoCustomTypeKind::ResourceAddress
+            | ScryptoCustomTypeKind::Own
+            | ScryptoCustomTypeKind::NonFungibleGlobalId
+            | ScryptoCustomTypeKind::Blob
+            | ScryptoCustomTypeKind::Bucket
+            | ScryptoCustomTypeKind::Proof
+            | ScryptoCustomTypeKind::Expression
+            | ScryptoCustomTypeKind::Hash
+            | ScryptoCustomTypeKind::EcdsaSecp256k1PublicKey
+            | ScryptoCustomTypeKind::EcdsaSecp256k1Signature
+            | ScryptoCustomTypeKind::EddsaEd25519PublicKey
+            | ScryptoCustomTypeKind::EddsaEd25519Signature
+            | ScryptoCustomTypeKind::Decimal
+            | ScryptoCustomTypeKind::PreciseDecimal
+            | ScryptoCustomTypeKind::NonFungibleLocalId
+            | ScryptoCustomTypeKind::KeyValueStore { .. } => {
+                validate_childless_metadata(type_metadata)?;
+            }
+        }
+        Ok(())
+    }
+
+    fn validate_type_validation_with_type_kind(
+        _: &TypeValidationContext,
+        type_kind: &SchemaCustomTypeKind<Self>,
+        _: &SchemaCustomTypeValidation<Self>,
+    ) -> Result<(), SchemaValidationError> {
+        // Even though they all map to the same thing, we keep the explicit match statement so that
+        // we will have to explicitly check this when we add a new `ScryptoCustomTypeKind`
+        match type_kind {
+            ScryptoCustomTypeKind::PackageAddress
+            | ScryptoCustomTypeKind::ComponentAddress
+            | ScryptoCustomTypeKind::ResourceAddress
+            | ScryptoCustomTypeKind::Own
+            | ScryptoCustomTypeKind::NonFungibleGlobalId
+            | ScryptoCustomTypeKind::Blob
+            | ScryptoCustomTypeKind::Bucket
+            | ScryptoCustomTypeKind::Proof
+            | ScryptoCustomTypeKind::Expression
+            | ScryptoCustomTypeKind::Hash
+            | ScryptoCustomTypeKind::EcdsaSecp256k1PublicKey
+            | ScryptoCustomTypeKind::EcdsaSecp256k1Signature
+            | ScryptoCustomTypeKind::EddsaEd25519PublicKey
+            | ScryptoCustomTypeKind::EddsaEd25519Signature
+            | ScryptoCustomTypeKind::Decimal
+            | ScryptoCustomTypeKind::PreciseDecimal
+            | ScryptoCustomTypeKind::NonFungibleLocalId
+            | ScryptoCustomTypeKind::KeyValueStore { .. } => {
+                // All other types currently should return early by having a `None` validation at the layer above,
+                // so if they get here then that's an error
+                return Err(SchemaValidationError::TypeValidationMismatch);
```

Comment:
Shouldn't this be a `panic!("Unreachable as there is no validation for Scrypto custom types")`

Reply:
Hmm no - we shouldn't panic here.

A Dapp can submit a manifest that claims to be a scrypto type kind, but uses a "String" validation which is invalid.
This means that their schema isn't valid, it doesn't mean that we should panic :D

Reply:
I'm confused here.

Doesn't the interface says `_: &SchemaCustomTypeValidation<Self>`. How come a `String` validation is used at the time this function is entered. If so, it should be checked by the layer above already.
```rust
        TypeKind::Custom(custom_type_kind) => {
            let TypeValidation::Custom(custom_type_validation) = type_validation else {
                return Err(SchemaValidationError::TypeValidationMismatch);
            };
            E::validate_type_validation_with_type_kind(
                context,
                custom_type_kind,
                custom_type_validation,
            )?;
        }
```


Maybe explictly match the input validation like
```rust
match validation {
   ScryptoCustomTypeValidation::None => {
           // No-op, no validation works for all scrypto custom types
   }
}
```

Reply:
All of these validation methods (for consistency) match on the type_kind and then cross-check on the type validation.

Right now, there is effectively no ScryptoCustomTypeValidation - in fact I've just made it an empty enum to be clear. So this method can never be called, and can be optimised away by the compiler.

BUT we will be adding them soon. So, I wrote the method to be forward-compatible.

BUT this confusion suggests to me that we should be passing in `SchemaTypeValidation<SchemaCustomTypeValidation<Self>>` instead perhaps?

Reply:
I've updated the comments and will merge. But happy to adjust if you think we should be passing the full set of validations down to the custom validator 👍

Reply:
In fact, possibly the custom validator should actually check any custom validation against any basic or custom type??

And the basic validator just checks the basic validations against the basic types. Maybe that's the better abstraction, as it allows the extension to add its own custom validations of basic types, which might be powerful?